packages/lesswrong/server/manualMigrations/2024-10-31-undoBestOfCollectionClobbering.ts

```diff
@@ -0,0 +1,155 @@
+  { postId: 'nNqXfnjiezYukiMJi', sequenceId: 'n945eovrA3oDueqtq' },
+  { postId: 'RQpNHSiWaXTvDxt6R', sequenceId: 'FYMiCeXEgMzsB5stm' },
+];
+
+registerMigration({
+  name: 'undoBestOfCollectionClobbering',
+  dateWritten: '2024-10-31',
+  idempotent: true,
+  action: async () => {
+    const collectionId = 'nmk3nLpQE89dMRzzN'; 
+
+    // Step 1: Remove canonical fields from affected posts
+    const collectionDoc = await Collections.findOne({ _id: collectionId });
+    if (!collectionDoc) {
+      // eslint-disable-next-line no-console
+      console.error(`Collection with ID ${collectionId} not found.`);
+      return;
+    }
+    const collectionSlug = collectionDoc.slug;
+
+    const postsToUnset = await Posts.find({ canonicalCollectionSlug: collectionSlug }).fetch();
+    // eslint-disable-next-line no-console
+    console.log(`Found ${postsToUnset.length} posts to unset canonical fields.`);
+
+    if (postsToUnset.length > 0) {
+      const updates = postsToUnset.map(post => ({
+        updateOne: {
+          filter: { _id: post._id },
+          update: {
+            $unset: {
+              canonicalPrevPostSlug: "",
+              canonicalNextPostSlug: "",
+              canonicalBookId: "",
+              canonicalCollectionSlug: "",
+              canonicalSequenceId: "",
+            }
+          }
+        }
+      }));
+      await Posts.rawCollection().bulkWrite(updates, { ordered: false });
+      // eslint-disable-next-line no-console
+      console.log(`Unset canonical fields for ${updates.length} posts.`);
+    }
+
+    // Step 2: Update canonical fields for specified posts
+    const postsBySequenceId = _.groupBy(POSTS_TO_UPDATE, 'sequenceId');
+    for (const sequenceId in postsBySequenceId) {
+      const postsList = postsBySequenceId[sequenceId];
+      const postIdsToUpdate = postsList.map(item => item.postId);
+
+      const sequence = await Sequences.findOne({ _id: sequenceId });
+      if (!sequence) {
+        // eslint-disable-next-line no-console
+        console.error(`Sequence with ID ${sequenceId} not found.`);
+        continue;
+      }
+
+      const postsInSequence = await getPostsInSequence(sequenceId);
+      const postsToUpdateInOrder = postsInSequence.filter(post => postIdsToUpdate.includes(post._id));
+
+      const updates = [];
+      for (let i = 0; i < postsToUpdateInOrder.length; i++) {
+        const currentPost = postsToUpdateInOrder[i];
+        const prevPost = i > 0 ? postsToUpdateInOrder[i - 1] : null;
+        const nextPost = i < postsToUpdateInOrder.length - 1 ? postsToUpdateInOrder[i + 1] : null;
+
+        updates.push({
+          updateOne: {
+            filter: { _id: currentPost._id },
+            update: {
+              $set: {
+                canonicalPrevPostSlug: prevPost ? prevPost.slug : '',
+                canonicalNextPostSlug: nextPost ? nextPost.slug : '',
+                canonicalSequenceId: sequenceId,
```

Comment:
Any chance these'll be missing their previous canonical book/collection IDs?

Reply:
These are all for user-made sequences, and users can't make collections (there are only 5) or books, so I don't think so?